### PR TITLE
Laboratory Work #1, Anna Sizykh - 20FPL3

### PR DESCRIPTION
**Pull request by AnnaSizykh**

### CODE REVIEW

File: seminars/03.04.2022/try_exceptions.py [hunk 1]
```diff
@@ -1,23 +1,34 @@
 # TASK 1
```
Comment (demid5111) [inline seminars/03.04.2022/try_exceptions.py:1]: discard the changes in unrelated files

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+requests
```
Comment (demid5111) [inline requirements.txt:1]: specify a particular version

File: scrapper.py [hunk 3]
```diff
@@ -29,6 +29,9 @@ def __init__(self, seed_urls, max_articles: int):
         pass
 
     def _extract_url(self, article_bs):
+        response = requests.get('https://periodical.pstgu.ru/ru/series/index/3')
```
Comment (demid5111) [inline scrapper.py:32]: this is an incorrect place for downloading the page. this function is responsible for extraction of standalone artile URL

**Review by demid5111 — CHANGES_REQUESTED** [review]
it is just a beginning

**Comment by demid5111** [comment]
there is no progress to this moment. please start the implementation process - there is plenty of actions to do!

File: scrapper.py [hunk 4]
```diff
@@ -1,6 +1,17 @@
 """
 Scrapper implementation
 """
+import json
```
Comment (demid5111) [inline scrapper.py:4]: all imports should be grouped into:

1. standard Python imports
2. 3rd party imports
3. project imports

all imports should be in alphabetic order

File: scrapper.py [hunk 5]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
     """
-    def __init__(self, seed_urls, max_articles: int):
-        pass
 
-    def _extract_url(self, article_bs):
-        pass
+    def __init__(self, seed_urls, max_articles: int):
+        self.max_articles = max_articles
+        self.seed_urls = seed_urls
+        self.count_articles = 0
```
Comment (demid5111) [inline scrapper.py:43]: that was not enumerated in allowed fields. remove

File: scrapper.py [hunk 6]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        self.urls = []
+
+    def _extract_url(self, article_bs):  # I don't understand if I need to collect seed urls (from main to year)
```
Comment (demid5111) [inline scrapper.py:46]: remove such questionable comments from the code and ask the question in the forum.

comments are needed for describing the code and not you concerns

File: scrapper.py [hunk 7]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        for article_code in articles_code_stored_here:
+            links_articles = article_code.find_all('a')
+            for link in links_articles:
+                if self.max_articles == self.count_articles:
+                    break
+                try:
+                    art_link = link["href"]
+                    if art_link.startswith('/') and 'pdf' not in art_link:
+                        self.urls.append('https://periodical.pstgu.ru' + link["href"])
+                        self.count_articles += 1
```
Comment (demid5111) [inline scrapper.py:58]: you might want to use a local variable instead?

File: scrapper.py [hunk 8]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
-        pass
+    def __init__(self, seed_urls, max_articles: int):
+        self.max_articles = max_articles
+        self.seed_urls = seed_urls
+        self.count_articles = 0
+        self.urls = []
+
+    def _extract_url(self, article_bs):  # I don't understand if I need to collect seed urls (from main to year)
+        # or just put it in scrapper_config.
+        articles_code_stored_here = article_bs.find_all('div', {'class': 'articles'})
```
Comment (demid5111) [inline scrapper.py:48]: remember to name variables pointing to bs object with `_bs` suffix

File: scrapper.py [hunk 9]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        for article_code in articles_code_stored_here:
+            links_articles = article_code.find_all('a')
+            for link in links_articles:
+                if self.max_articles == self.count_articles:
+                    break
+                try:
+                    art_link = link["href"]
+                    if art_link.startswith('/') and 'pdf' not in art_link:
+                        self.urls.append('https://periodical.pstgu.ru' + link["href"])
```
Comment (demid5111) [inline scrapper.py:57]: there should be a better place for a string constant that represents hostname of your website

File: scrapper.py [hunk 10]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+
+    def _extract_url(self, article_bs):  # I don't understand if I need to collect seed urls (from main to year)
+        # or just put it in scrapper_config.
+        articles_code_stored_here = article_bs.find_all('div', {'class': 'articles'})
+        for article_code in articles_code_stored_here:
+            links_articles = article_code.find_all('a')
+            for link in links_articles:
+                if self.max_articles == self.count_articles:
+                    break
+                try:
```
Comment (demid5111) [inline scrapper.py:54]: too big try/except. it should surround only vulnerable piece of code

File: scrapper.py [hunk 11]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+                        self.count_articles += 1
+                except KeyError:
+                    continue
 
     def find_articles(self):
         """
         Finds articles
-        """
-        pass
+        """  # I also don't understand what I need to do here, because this code can be moved to extract_url
```
Comment (demid5111) [inline scrapper.py:65]: ?

File: scrapper.py [hunk 12]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+    def __init__(self, seed_urls, max_articles: int):
+        self.max_articles = max_articles
+        self.seed_urls = seed_urls
+        self.count_articles = 0
+        self.urls = []
+
+    def _extract_url(self, article_bs):  # I don't understand if I need to collect seed urls (from main to year)
+        # or just put it in scrapper_config.
+        articles_code_stored_here = article_bs.find_all('div', {'class': 'articles'})
+        for article_code in articles_code_stored_here:
```
Comment (demid5111) [inline scrapper.py:49]: extract url is nor responsible for controlling number of URLs that were already collected versus the required number of URLs

File: scrapper.py [hunk 13]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+                    continue
 
     def find_articles(self):
         """
         Finds articles
-        """
-        pass
+        """  # I also don't understand what I need to do here, because this code can be moved to extract_url
+        for seed_url in self.seed_urls:
+            issue_page = requests.get(seed_url)
```
Comment (demid5111) [inline scrapper.py:67]: any attempts to make the remote server request more human like? do you recall several strategies that wllo Python requests look like human-made?

File: scrapper.py [hunk 14]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        return self.article
+
+    def _fill_article_with_text(self, article_bs):
+        list_of_possible_l = ['Список литературы', 'ЛИТЕРАТУРА', 'Список использованных источников и литературы',
+                              'Источники']
+
+        class_of_link = article_bs.find('div', {'class': "articles-item-more"})
+        if class_of_link:
+            link_to_save = class_of_link.find('a')
+            link_to_pdf = 'https://periodical.pstgu.ru' + link_to_save['href']
```
Comment (demid5111) [inline scrapper.py:98]: do you see - you again use that constant! another reason to think of central place for defining it

File: scrapper.py [hunk 15]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+            final_page = requests.get(link_to_pdf)
+            final_page_bs = BeautifulSoup(final_page.text, 'html.parser')
+        else:
+            final_page_bs = article_bs
```
Comment (demid5111) [inline scrapper.py:102]: why do we need this condition?

File: scrapper.py [hunk 16]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        url_pattern = re.compile(r"var DEFAULT_URL = '(.*?)'")
+        final_link = final_page_bs.find(string=url_pattern)
+        final_link = 'https://periodical.pstgu.ru' + \
```
Comment (demid5111) [inline scrapper.py:105]: 3rd place of using this constant

File: scrapper.py [hunk 17]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+                     str(final_link).replace('var DEFAULT_URL = ', '').replace("'", '').replace(';', '').strip()
```
Comment (demid5111) [inline scrapper.py:106]: I imagine how it difficult is was 👏

File: scrapper.py [hunk 18]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        raw_pdf = PDFRawFile(final_link, self.article_id)
+        raw_pdf.download()
+        text_from_pdf = raw_pdf.get_text()
+        for refs_maker in list_of_possible_l:
+            if refs_maker in text_from_pdf:
+                text_pdf_list = text_from_pdf.split(refs_maker)
+                del text_pdf_list[-1]
```
Comment (demid5111) [inline scrapper.py:113]: ??

File: scrapper.py [hunk 19]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+            self.article.author = author.text
+        except AttributeError:
+            self.article.author = 'No data'
+        title = article_bs.find('h2', {"class": "articles-item-name"})
+        self.article.title = title.text
+        topics = article_bs.find('div', {"class": "articles"}).find_all('p')
+        try:
+            self.article.topics = topics[2].text
+        except IndexError:
+            self.article.topics = 'No data'
```
Comment (demid5111) [inline scrapper.py:131]: that is not correct, consult the docs. I am sure you already know it thanks to tests

File: scrapper.py [hunk 20]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        year = article_bs.find('meta', {'name': 'citation_publication_date'})['content']
+        issue_title = article_bs.find('meta', {'name': 'citation_journal_title'})['content']
+        for letter in issue_title:
+            if letter.isdigit():
+                issue_number.append(letter)
```
Comment (demid5111) [inline scrapper.py:136]: ?? can you propose a better way for doing that?

File: scrapper.py [hunk 21]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        month = month_dict[int(''.join(issue_number)) % 4]
+        text_date = month + '.' + year
```
Comment (demid5111) [inline scrapper.py:138]: use f-string instead

File: scrapper.py [hunk 22]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        date = datetime.datetime.strptime(text_date, '%m.%Y')
+        self.article.date = date
 
 
 def prepare_environment(base_path):
     """
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    base_path_p = pathlib.Path(base_path)
```
Comment (demid5111) [inline scrapper.py:147]: refactor to use pathlib. os.path is deprecated for these tasks

File: scrapper.py [hunk 23]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
 
 
 def validate_config(crawler_path):
     """
     Validates given config
     """
-    pass
+    with open(crawler_path) as path_thing_idk:
+        config_dict = json.load(path_thing_idk)
+    seed_urls = config_dict['seed_urls']
```
Comment (demid5111) [inline scrapper.py:161]: are you sure those keys will be always present in such a config? will you accept an incomplete config?

File: scrapper.py [hunk 24]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+        if not seed_url.startswith('https://'):
+            raise IncorrectURLError
+
+    return seed_urls, max_articles
 
 
 if __name__ == '__main__':
     # YOUR CODE HERE
-    pass
+    i = 0
```
Comment (demid5111) [inline scrapper.py:182]: initialize closer to the loop

File: scrapper.py [hunk 25]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+    s_urls, max_as = validate_config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(s_urls, max_as)
+    crawler.find_articles()
+    print(len(crawler.urls))
```
Comment (demid5111) [inline scrapper.py:187]: remove debug print

File: scrapper.py [hunk 26]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
+    for art_url in crawler.urls:
+        i += 1
+        print(i)
```
Comment (demid5111) [inline scrapper.py:190]: ?

File: scrapper.py [hunk 27]
```diff
@@ -25,39 +36,159 @@ class Crawler:
...
 if __name__ == '__main__':
     # YOUR CODE HERE
-    pass
+    i = 0
+    s_urls, max_as = validate_config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(s_urls, max_as)
+    crawler.find_articles()
+    print(len(crawler.urls))
+    for art_url in crawler.urls:
```
Comment (demid5111) [inline scrapper.py:188]: there should be a more elegant solution for getting access to the index of current element in the loop

File: scrapper_config.json [hunk 28]
```diff
@@ -1,4 +1,6 @@
 {
-    "seed_urls": [],
-    "total_articles_to_find_and_parse": 0
+    "seed_urls": ["https://periodical.pstgu.ru/ru/series/issue/3/69",
+        "https://periodical.pstgu.ru/ru/series/issue/3/68", "https://periodical.pstgu.ru/ru/series/issue/3/67",
+        "https://periodical.pstgu.ru/ru/series/issue/3/66"],
+    "total_articles_to_find_and_parse": 15
```
Comment (demid5111) [inline scrapper_config.json:5]: have you checked for collecting 100 articles?

**Review by demid5111 — COMMENTED** [review]
progress is visible, but there are some serios things to consider and also align with docs

File: scrapper.py (hunk 18 above)
Comment (AnnaSizykh) [reply scrapper.py:113]: I used this part of the code to separate and delete the list of references ( as it was a part of the task).

File: scrapper.py [hunk 29]
```diff
@@ -1,6 +1,19 @@
 """
 Scrapper implementation
 """
+import datetime
+import json
+import os
```
Comment (demid5111) [inline scrapper.py:6]: use pathlib

File: scrapper.py [hunk 30]
```diff
@@ -26,38 +39,156 @@ class Crawler:
...
+    if max_articles > 150:
+        raise NumberOfArticlesOutOfRangeError
+    if not isinstance(seed_urls, list):
+        raise IncorrectURLError
+    if len(seed_urls) <= 1:
+        raise IncorrectURLError
+    for seed_url in seed_urls:
+        if not isinstance(seed_url, str):
+            raise IncorrectURLError
+        if not seed_url.startswith('https://'):
```
Comment (demid5111) [inline scrapper.py:175]: https://github.com/fipl-hse/2021-2-level-ctlr/pull/25/files

File: scrapper.py [hunk 31]
```diff
@@ -26,38 +39,156 @@ class Crawler:
...
 
 
 def validate_config(crawler_path):
     """
     Validates given config
     """
-    pass
+    with open(crawler_path) as path_thing_idk:
+        config_dict = json.load(path_thing_idk)
+    try:
```
Comment (demid5111) [inline scrapper.py:158]: easier way

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Review by demid5111 — COMMENTED** [review]
start working on text processing pipeline, deadline will approach shortly

**Review by demid5111 — CHANGES_REQUESTED** [review]
no progress since last time!

**Comment by demid5111** [comment]
no progress with a pipeline, hurry up!

File: config/target_score.txt [hunk 32]
```diff
@@ -1,5 +1,5 @@
 # Target score for scrapper.py:
-4
+8
 
 # Target score for pipeline.py:
-0
+4
```
Comment (demid5111) [inline config/target_score.txt]: too small, I expect a much higher mark!

File: pipeline.py [hunk 33]
```diff
@@ -51,19 +61,26 @@ class CorpusManager:
...
+        self._storage = {}
+        self.path_to_raw_txt_data = path_to_raw_txt_data
+        self._scan_dataset()
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        paths_to_files = list(Path(self.path_to_raw_txt_data).glob('*_raw.txt'))
```
Comment (demid5111) [inline pipeline.py]: why do you create a list - instead of directly iterating through a glob result? that would be much less memory consuming

File: pipeline.py [hunk 34]
```diff
@@ -51,19 +61,26 @@ class CorpusManager:
...
+        id_template = re.compile(r'\d+')
+        for file_path in paths_to_files:
+            file_id = int(re.search(id_template, str(file_path.name)).group())
```
Comment (demid5111) [inline pipeline.py]: as soon as you have a compiled pattern, you need to call methods from it, like: `id_template.search`

Comment (demid5111) [inline pipeline.py]: is not a file_path.name already a string?

File: pipeline.py [hunk 35]
```diff
@@ -72,31 +89,79 @@ class TextProcessingPipeline:
...
         """
         Runs pipeline process scenario
         """
-        pass
+        for article_item in self.corpus_manager.get_articles().values():
+            article_text = article_item.get_raw_text()
+            article_text = article_text.replace('-\n', '')
+            for symbol in ['\n', '\r']:
+                article_text = article_text.replace(symbol, ' ')
+            re_tokens = re.findall(r'[а-яё_]+-?[а-яё_]*', article_text, flags=re.IGNORECASE)
```
Comment (demid5111) [inline pipeline.py]: again, pattern in a loop

File: pipeline.py [hunk 36]
```diff
@@ -72,31 +89,79 @@ class TextProcessingPipeline:
...
         pass
 
     def run(self):
         """
         Runs pipeline process scenario
         """
-        pass
+        for article_item in self.corpus_manager.get_articles().values():
+            article_text = article_item.get_raw_text()
+            article_text = article_text.replace('-\n', '')
```
Comment (demid5111) [inline pipeline.py]: you should do cleaning with mystem as well

File: pipeline.py [hunk 37]
```diff
@@ -72,31 +89,79 @@ class TextProcessingPipeline:
...
+            for symbol in ['\n', '\r']:
+                article_text = article_text.replace(symbol, ' ')
+            re_tokens = re.findall(r'[а-яё_]+-?[а-яё_]*', article_text, flags=re.IGNORECASE)
+            cleaned_text = ' '.join(re_tokens)
+            article_item.save_as(cleaned_text.lower(), 'cleaned')
+            self._process(cleaned_text)
```
Comment (demid5111) [inline pipeline.py]: process should return the list of morphological tokens

File: pipeline.py [hunk 38]
```diff
@@ -72,31 +89,79 @@ class TextProcessingPipeline:
...
         """
-        pass
+        for article_item in self.corpus_manager.get_articles().values():
+            article_text = article_item.get_raw_text()
+            article_text = article_text.replace('-\n', '')
+            for symbol in ['\n', '\r']:
+                article_text = article_text.replace(symbol, ' ')
+            re_tokens = re.findall(r'[а-яё_]+-?[а-яё_]*', article_text, flags=re.IGNORECASE)
+            cleaned_text = ' '.join(re_tokens)
+            article_item.save_as(cleaned_text.lower(), 'cleaned')
```
Comment (demid5111) [inline pipeline.py]: save of article should happen after you run processing. Otherwise it breaks the overall logic

File: pipeline.py [hunk 39]
```diff
@@ -72,31 +89,79 @@ class TextProcessingPipeline:
...
+        raise InconsistentDatasetError
+    file_names = []
+    for files_path in children_files:
+        file_names.append(files_path.name)
+    if len(file_names) % 2 != 0:
+        raise InconsistentDatasetError
+    for i in range(1, int(len(list(children_files)) / 2) + 1):
+        if (f'{i}_raw.txt' not in file_names) or (f'{i}_meta.json' not in file_names):
+            raise FileNotFoundError
+    for file in children_files:
```
Comment (demid5111) [inline pipeline.py:160]: what is the reason to open the file if it exists?

File: pipeline.py [hunk 40]
```diff
@@ -72,31 +89,79 @@ class TextProcessingPipeline:
...
         Processes each token and creates MorphToken class instance
         """
-        pass
+        m_tokens_list = []
+        analyzed_text_mystem = Mystem().analyze(raw_text)
+        for analyzed_word in analyzed_text_mystem:
+            if analyzed_word['text'] == ' ':
+                continue
+            original_word = analyzed_word['text']
+            token = MorphologicalToken(original_word)
```
Comment (demid5111) [inline pipeline.py:127]: what is the reason to create an instance before making the checks? if checks fail, this instance is not used

**Review by demid5111 — CHANGES_REQUESTED** [review]
I see how you tackle problems that are always present, nice progress!

File: pipeline.py (hunk 39 above)
Comment (AnnaSizykh) [reply pipeline.py:160]: To check if the file is empty or not? I believe we should have had a test for it. Do you want me to do it some other way?

File: config/target_score.txt [hunk 41]
```diff
@@ -1,5 +1,5 @@
 # Target score for scrapper.py:
-4
+8
 
 # Target score for pipeline.py:
-0
+10
```
Comment (demid5111) [inline config/target_score.txt:5]: awesome!

File: pipeline.py [hunk 42]
```diff
@@ -72,31 +90,89 @@ class TextProcessingPipeline:
...
+            article_item.save_as(' '.join(multiple_tagged_tokens), kind='multiple_tagged')
 
     def _process(self, raw_text: str):
         """
         Processes each token and creates MorphToken class instance
         """
-        pass
+        article_text = raw_text.replace('-\n', '')
+        for symbol in ['\n', '\r']:
+            article_text = article_text.replace(symbol, ' ')
```
Comment (demid5111) [inline pipeline.py]: I do not see a reason to make such a preprocessing. Mystem can accept "dirty" text (with all the punctuations and other staff) and then it will process it accordingly, you only need to properly parse its output report

File: pipeline.py [hunk 43]
```diff
@@ -72,31 +90,89 @@ class TextProcessingPipeline:
...
+            if not analyzed_word['analysis']:
+                continue
+            if 'lex' not in analyzed_word['analysis'][0].keys():
+                continue
+            if 'gr' not in analyzed_word['analysis'][0].keys():
+                continue
+            token = MorphologicalToken(original_word)
+            token.normalized_form = analyzed_word['analysis'][0].get('lex')
+            token.tags_mystem = analyzed_word['analysis'][0].get('gr')
+            m_tokens_list.append(token)
```
Comment (demid5111) [inline pipeline.py:134]: a cleaner solution would be append only after you fill it, otherwise it is hard to follow

File: pipeline.py [hunk 44]
```diff
@@ -72,31 +90,89 @@ class TextProcessingPipeline:
...
+        for symbol in ['\n', '\r']:
+            article_text = article_text.replace(symbol, ' ')
+        m_tokens_list = []
+        analyzed_text_mystem = Mystem().analyze(article_text)
+        pymorphy2_analyzer = pymorphy2.MorphAnalyzer()
+        for analyzed_word in analyzed_text_mystem:
+            if analyzed_word['text'] == ' ':
+                continue
+            original_word = analyzed_word['text']
+            if 'analysis' not in analyzed_word.keys():
```
Comment (demid5111) [inline pipeline.py]: could we join the conditions somehow?

File: pipeline.py (hunk 39 above)
Comment (demid5111) [reply pipeline.py:160]: where did you find a requirement to open the file for emptiness? If this is not stated in docs, remove. Otherwise, please point to that place

Comment (demid5111) [reply pipeline.py:160]: but if you insist, there is a much smarter check that file is empty. How do you think what am I talking about?

File: pipeline.py [hunk 45]
```diff
@@ -72,31 +90,89 @@ class TextProcessingPipeline:
...
+    if not path_to_validate.is_dir():
+        raise NotADirectoryError
+    children_files = list(path_to_validate.glob('*raw.txt'))
+    children_files.extend(list(path_to_validate.glob('*.json')))
+    if not children_files:
+        raise EmptyDirectoryError
+    file_names = []
+    for files_path in children_files:
+        file_names.append(files_path.name)
+    if len(file_names) % 2 != 0:
```
Comment (demid5111) [inline pipeline.py]: very naive check

File: pos_frequency_pipeline.py [hunk 46]
```diff
@@ -13,15 +18,41 @@ class EmptyFileError(Exception):
...
+        for article_item in self.corpus_manager.get_articles().values():
+            article_tag_path = article_item.get_file_path('single_tagged')
+            frequencies_dict = {}
+            pos_pattern = re.compile(r'[A-Z]+')
+            with open(article_tag_path, 'r', encoding='utf-8') as text:
+                text_n_tags = text.read()
+                if not text_n_tags:
+                    raise EmptyFileError
+            pos_all = pos_pattern.findall(text_n_tags)
+            print(len(pos_all))
```
Comment (demid5111) [inline pos_frequency_pipeline.py]: rm

File: pos_frequency_pipeline.py [hunk 47]
```diff
@@ -13,15 +18,41 @@ class EmptyFileError(Exception):
...
+            for pos in pos_all:
+                if pos not in frequencies_dict:
+                    frequencies_dict[pos] = pos_all.count(pos)
+            visualize(statistics=frequencies_dict,
```
Comment (demid5111) [inline pos_frequency_pipeline.py]: change the order: first save json, second visualize

File: requirements_qa.txt [hunk 48]
```diff
@@ -1,6 +1,6 @@
 ghapi==0.1.19
-matplotlib==3.3.4
-numpy==1.20.1
+matplotlib
```
Comment (demid5111) [inline requirements_qa.txt]: discard changes here

**Review by demid5111 — CHANGES_REQUESTED** [review]
only some minor comments left

File: pipeline.py [hunk 49]
```diff
@@ -51,19 +62,26 @@ class CorpusManager:
...
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        paths_to_files = Path(self.path_to_raw_txt_data).glob('*_raw.txt')
+        id_template = re.compile(r'\d+')
+        for file_path in paths_to_files:
+            file_id = int(id_template.search(file_path.name).group())
```
Comment (demid5111) [inline pipeline.py]: number is not guaranteed

File: pos_frequency_pipeline.py [hunk 50]
```diff
@@ -13,15 +18,38 @@ class EmptyFileError(Exception):
...
     def __init__(self, corpus_manager: CorpusManager):
-        pass
+        self.corpus_manager = corpus_manager
 
     def run(self):
-        pass
+        for article_item in self.corpus_manager.get_articles().values():
+            article_tag_path = article_item.get_file_path('single_tagged')
+            frequencies_dict = {}
+            pos_pattern = re.compile(r'[A-Z]+')
```
Comment (demid5111) [inline pos_frequency_pipeline.py]: more precise

**Review by demid5111 — CHANGES_REQUESTED** [review]
